pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,466 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+
+from sumpy.symbolic import make_sym_vector, SympyToPymbolicMapper
+import sumpy.symbolic as sym
+from sumpy.kernel import (AxisTargetDerivative, AxisSourceDerivative,
+    ExpressionKernel, KernelWrapper, TargetPointMultiplier)
+from pytools import (memoize_on_first_arg,
+                generate_nonnegative_integer_tuples_summing_to_at_most
+                as gnitstam)
+
+from pytential.symbolic.primitives import (NodeCoordinateComponent,
+    hashable_kernel_args)
+from pytential.symbolic.mappers import IdentityMapper
+from pytential.utils import chop, lu_solve_with_expand
+import pytential
+
+import logging
+logger = logging.getLogger(__name__)
+
+__all__ = (
+    "rewrite_using_base_kernel",
+    "get_deriv_relation",
+    )
+
+__doc__ = """
+.. autofunction:: rewrite_using_base_kernel
+.. autofunction:: get_deriv_relation
+"""
+
+
+# {{{ rewrite_using_base_kernel
+
+_NO_ARG_SENTINEL = object()
+
+
+def rewrite_using_base_kernel(exprs, base_kernel=_NO_ARG_SENTINEL):
+    """Rewrites an expression with :class:`~pytential.symbolic.primitives.IntG`
+    objects using *base_kernel*.
+
+    For example, if *base_kernel* is the Biharmonic kernel, and a Laplace kernel
+    is encountered, this will (forcibly) rewrite the Laplace kernel in terms of
+    derivatives of the Biharmonic kernel.
+
+    The routine will fail if this process cannot be completed.
+    """
+    if base_kernel is None:
+        return list(exprs)
+    mapper = RewriteUsingBaseKernelMapper(base_kernel)
+    return [mapper(expr) for expr in exprs]
+
+
+class RewriteUsingBaseKernelMapper(IdentityMapper):
+    """Rewrites IntGs using the base kernel. First this method replaces
+    IntGs with :class:`sumpy.kernel.AxisTargetDerivative` to IntGs
+    :class:`sumpy.kernel.AxisSourceDerivative` and then replaces
+    IntGs with :class:`sumpy.kernel.TargetPointMultiplier` to IntGs
+    without them using :class:`sumpy.kernel.ExpressionKernel`
+    and then finally converts them to the base kernel by finding
+    a relationship between the derivatives.
+    """
+    def __init__(self, base_kernel):
+        self.base_kernel = base_kernel
+
+    def map_int_g(self, expr):
+        # First convert IntGs with target derivatives to source derivatives
+        expr = convert_target_deriv_to_source(expr)
+        # Convert IntGs with TargetMultiplier to a sum of IntGs without
+        # TargetMultipliers
+        new_int_gs = convert_target_multiplier_to_source(expr)
+        # Convert IntGs with different kernels to expressions containing
+        # IntGs with base_kernel or its derivatives
+        return sum(convert_int_g_to_base(new_int_g,
+            self.base_kernel) for new_int_g in new_int_gs)
```

Comment:
Do each of these have unit tests? :duck:

Reply:
Added a new test file `test/test_system_pde_utils.py` with unit tests